### ISSUE TITLE
`bash: line 1: main: command not found` when the disk is full

### ISSUE THREAD

**Issue by gleniat** [dokku/dokku#5764]
### Description of problem

Dokku displays the message `bash: line 1: main: command not found` when the disk is full, instead of providing a more helpful error message.

### Steps to reproduce

1. Gradually fill the disk space by regularly using Dokku, deploying apps, and pushing code, without manually scheduling a daily Docker prune to prevent the disk from filling up. Alternatively, use the 'dd' command to rapidly fill up the disk.

2. After the disk is full, execute any Dokku command.

```
$ dokku ps:restart mywebapp
 !     /home/dokku/.basher/bash: line 1: main: command not found
2023/03/20 08:13:22 exit status 127
```

```
$ dokku ps:stop mywebapp
 !     /home/dokku/.basher/bash: line 1: main: command not found
2023/03/20 08:13:28 exit status 127
```

```
$ dokku app:destroy mywebapp
 !     /home/dokku/.basher/bash: line 1: main: command not found
2023/03/20 08:14:16 exit status 127
```

### dokku report $APP_NAME

N/A

**Comment by josegonzalez** [comment]
If you still have a full disk, mind doing a `dokku --trace app:destroy mywebapp` so I can see exactly where it fails?

**Comment by josegonzalez** [comment]
Closing this for now. If someone can replicate and add a `--trace` to the command, that would be _awesome_ :)

**Comment by boozec** [comment]
Hi @josegonzalez I've got the same error. Here a trace as you suggested:

```
+ SSH_USER=root
+ export SSH_NAME=default
+ SSH_NAME=default
+ export DOKKU_COMMAND=apps:destroy
+ DOKKU_COMMAND=apps:destroy
++ find /var/lib/dokku/plugins/enabled/20_events/user-auth
++ wc -l
+ local user_auth_count=1
+ [[ 1 == 0 ]]
+ [[ 1 == 1 ]]
+ [[ -f /var/lib/dokku/plugins/enabled/20_events/user-auth ]]
+ return 0
+ case "$1" in
+ execute_dokku_cmd apps:destroy NAME
+ declare 'desc=executes dokku sub-commands'
+ local PLUGIN_NAME=apps:destroy
+ local PLUGIN_CMD=apps:destroy
+ local implemented=0
+ local script
+ argv=('apps:destroy' 'NAME')
+ local argv
+ case "$PLUGIN_NAME" in
++ readlink -f /var/lib/dokku/plugins/enabled/apps
+ [[ /var/lib/dokku/core-plugins/available/apps == *core-plugins* ]]
+ [[ apps:destroy == \a\p\p\s\:\d\e\s\t\r\o\y ]]
+ shift 1
+ [[ -n '' ]]
+ set -- apps:destroy NAME
+ [[ -x /var/lib/dokku/plugins/enabled/apps:destroy/subcommands/default ]]
+ [[ -x /var/lib/dokku/plugins/enabled/apps:destroy/subcommands/apps:destroy ]]
+ [[ -x /var/lib/dokku/plugins/enabled/apps/subcommands/destroy ]]
+ [[ -n destroy ]]
+ /var/lib/dokku/plugins/enabled/apps/subcommands/destroy apps:destroy NAME
 !     WARNING: Potentially Destructive Action
 !     This command will destroy app NAME.
 !     To proceed, type "NAME"
> NAME
-----> Destroying NAME (including all add-ons)
 ?     plugn trigger deployed-app-image-tag [NAME]
 ?     plugn trigger deployed-app-image-tag stderr: /home/dokku/.basher/bash: line 1: main: command not found
 ?     plugn trigger deployed-app-image-tag stderr:
 ?     plugn trigger deployed-app-image-tag stdout:
 ?     plugn trigger pre-delete [NAME ]
/home/dokku/.basher/bash: line 1: main: command not found
 !     exit status 127
```

**Comment by shehaj** [comment]
Hello @josegonzalez , I am facing also this issue, here is the trace:
```
+ export DOKKU_HOST_ROOT=/home/dokku
+ DOKKU_HOST_ROOT=/home/dokku
+ export DOKKU_DISTRO
++ . /etc/os-release
++ echo debian
+ DOKKU_DISTRO=debian
+ export DOCKER_BIN=docker
+ DOCKER_BIN=docker
+ export DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ export DOKKU_CNB_BUILDER=heroku/builder:22
+ DOKKU_CNB_BUILDER=heroku/builder:22
+ export DOKKU_LIB_ROOT=/var/lib/dokku
+ DOKKU_LIB_ROOT=/var/lib/dokku
+ export PLUGIN_PATH=/var/lib/dokku/plugins
+ PLUGIN_PATH=/var/lib/dokku/plugins
+ export PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ export PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ export PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ export PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ export PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ export DOKKU_SYSTEM_GROUP=dokku
+ DOKKU_SYSTEM_GROUP=dokku
+ export DOKKU_SYSTEM_USER=dokku
+ DOKKU_SYSTEM_USER=dokku
+ export DOKKU_API_VERSION=1
+ DOKKU_API_VERSION=1
+ export DOKKU_NOT_IMPLEMENTED_EXIT=10
+ DOKKU_NOT_IMPLEMENTED_EXIT=10
+ export DOKKU_VALID_EXIT=0
+ DOKKU_VALID_EXIT=0
+ export DOKKU_PID=1859555
+ DOKKU_PID=1859555
+ export DOKKU_LOGS_DIR=/var/log/dokku
+ DOKKU_LOGS_DIR=/var/log/dokku
+ export DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ export DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ export DOKKU_CONTAINER_LABEL=dokku
+ DOKKU_CONTAINER_LABEL=dokku
+ export 'DOKKU_GLOBAL_BUILD_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_BUILD_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ export 'DOKKU_GLOBAL_RUN_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_RUN_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ parse_args apps:destroy facebook-ads-frontend
+ declare 'desc=top-level cli arg parser'
+ local next_index=1
+ local skip=false
+ args=('apps:destroy' 'facebook-ads-frontend')
+ local args
+ local flags
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ apps:destroy == \-\-\a\p\p ]]
+ [[ apps:destroy =~ ^--.* ]]
+ next_index=2
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ facebook-ads-frontend == \-\-\a\p\p ]]
+ [[ facebook-ads-frontend =~ ^--.* ]]
+ next_index=3
+ [[ -z --trace ]]
+ return 0
+ args=("$@")
+ skip_arg=false
+ [[ apps:destroy =~ ^--.* ]]
+ has_tty
+ declare 'desc=return 0 if we have a tty'
+ [[ '' == \t\r\u\e ]]
++ LC_ALL=C
++ /usr/bin/tty
+ [[ /dev/pts/0 == \n\o\t\ \a\ \t\t\y ]]
+ return 0
++ id -un
+ [[ dokku != \d\o\k\k\u ]]
+ [[ apps:destroy =~ ^plugin:.* ]]
+ [[ apps:destroy == \s\s\h\-\k\e\y\s\:\a\d\d ]]
+ [[ apps:destroy == \s\s\h\-\k\e\y\s\:\r\e\m\o\v\e ]]
+ [[ -n '' ]]
+ dokku_auth apps:destroy facebook-ads-frontend
+ declare 'desc=calls user-auth plugin trigger'
+ export SSH_USER=damianoshehaj97
+ SSH_USER=damianoshehaj97
+ export SSH_NAME=default
+ SSH_NAME=default
+ export DOKKU_COMMAND=apps:destroy
+ DOKKU_COMMAND=apps:destroy
++ wc -l
++ find /var/lib/dokku/plugins/enabled/20_events/user-auth
+ local user_auth_count=1
+ [[ 1 == 0 ]]
+ [[ 1 == 1 ]]
+ [[ -f /var/lib/dokku/plugins/enabled/20_events/user-auth ]]
+ return 0
+ case "$1" in
+ execute_dokku_cmd apps:destroy facebook-ads-frontend
+ declare 'desc=executes dokku sub-commands'
+ local PLUGIN_NAME=apps:destroy
+ local PLUGIN_CMD=apps:destroy
+ local implemented=0
+ local script
+ argv=('apps:destroy' 'facebook-ads-frontend')
+ local argv
+ case "$PLUGIN_NAME" in
++ readlink -f /var/lib/dokku/plugins/enabled/apps
+ [[ /var/lib/dokku/core-plugins/available/apps == *core-plugins* ]]
+ [[ apps:destroy == \a\p\p\s\:\d\e\s\t\r\o\y ]]
+ shift 1
+ [[ -n '' ]]
+ set -- apps:destroy facebook-ads-frontend
+ [[ -x /var/lib/dokku/plugins/enabled/apps:destroy/subcommands/default ]]
+ [[ -x /var/lib/dokku/plugins/enabled/apps:destroy/subcommands/apps:destroy ]]
+ [[ -x /var/lib/dokku/plugins/enabled/apps/subcommands/destroy ]]
+ [[ -n destroy ]]
+ /var/lib/dokku/plugins/enabled/apps/subcommands/destroy apps:destroy facebook-ads-frontend
 !     WARNING: Potentially Destructive Action
 !     This command will destroy app facebook-ads-frontend.
 !     To proceed, type "facebook-ads-frontend"
> facebook-ads-frontend
-----> Destroying facebook-ads-frontend (including all add-ons)
 ?     plugn trigger deployed-app-image-tag [facebook-ads-frontend]
 ?     plugn trigger deployed-app-image-tag stderr: /home/dokku/.basher/bash: line 1: main: command not found
 ?     plugn trigger deployed-app-image-tag stderr:
 ?     plugn trigger deployed-app-image-tag stdout:
 ?     plugn trigger pre-delete [facebook-ads-frontend ]
/home/dokku/.basher/bash: line 1: main: command not found
 !     exit status 127
```

**Comment by josegonzalez** [comment]
@shehaj if you still have a full disk, try running `docker system prune -a` to clear up space so basher can run. If it still doesn't work, please file a new ticket.

**Comment by gleniat** [comment]
@josegonzalez This report was not about us not being able to clear space on the disk, but about the fact that the error message `command not found` does not describe the real problem, and thus is not helpful. A message like "Not enough space on disk", or "Not enough space on disk, dokku commands may fail" may be more useful.

**Comment by josegonzalez** [comment]
Thanks for clarifying, was trying to ensure you were seeing the same issue and your problem was related.

**Comment by josegonzalez** [comment]
Seems like the problem is here: https://github.com/progrium/go-basher/blob/master/basher.go#L225

Plugn writes the file but doesn't validate that the file was actually closed and written, so the status of the file is lost in the ether.

What we should do is attempt to sync the file to disk, and fail if it can't. I believe the write failure should show more context than this. We can also show the free space in the `dokku report` output:

```shell
# disk space
df --type btrfs --type ext4 --type ext3 --type ext2 --type vfat --type iso9660

# inodes
df -hi --type btrfs --type ext4 --type ext3 --type ext2 --type vfat --type iso9660
```

While the error might not exactly say "disk is full", I think a better error from the write command + the df output in the `dokku report` output will help significantly.

**Comment by josegonzalez** [comment]
Okay I filed a ticket upstream, so I'll be able to track that separately from the report improvement here.

**Comment by josegonzalez** [comment]
Closing as there is a pull request open for the dokku side. I'll get to the plugn bug after the next Dokku minor (which should be soon).

Thanks for re-raising the issue :)

**Comment by gleniat** [comment]
Hi, @josegonzalez 
> If someone can replicate and add a `--trace` to the command, that would be _awesome_ :)
here is the trace for version 0.34.4.

```
# dokku --version
dokku version 0.34.4
```

```
# dokku ps:inspect myapp
/home/dokku/.basher/bash: line 1: main: command not found
 !     /home/dokku/.basher/bash: line 1: main: command not found
```

```
myserver:~# dokku ps:restart myapp
 !     /home/dokku/.basher/bash: line 1: main: command not found
```

```
myserver:~# dokku ps:inspect myapp --trace
+ export DOKKU_HOST_ROOT=/home/dokku
+ DOKKU_HOST_ROOT=/home/dokku
+ export DOKKU_DISTRO
++ . /etc/os-release
++ echo debian
+ DOKKU_DISTRO=debian
+ export DOCKER_BIN=docker
+ DOCKER_BIN=docker
+ export DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ export DOKKU_CNB_BUILDER=heroku/builder:22
+ DOKKU_CNB_BUILDER=heroku/builder:22
+ export DOKKU_LIB_ROOT=/var/lib/dokku
+ DOKKU_LIB_ROOT=/var/lib/dokku
+ export PLUGIN_PATH=/var/lib/dokku/plugins
+ PLUGIN_PATH=/var/lib/dokku/plugins
+ export PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ export PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ export PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ export PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ export PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ export DOKKU_SYSTEM_GROUP=dokku
+ DOKKU_SYSTEM_GROUP=dokku
+ export DOKKU_SYSTEM_USER=dokku
+ DOKKU_SYSTEM_USER=dokku
+ export DOKKU_API_VERSION=1
+ DOKKU_API_VERSION=1
+ export DOKKU_NOT_IMPLEMENTED_EXIT=10
+ DOKKU_NOT_IMPLEMENTED_EXIT=10
+ export DOKKU_VALID_EXIT=0
+ DOKKU_VALID_EXIT=0
+ export DOKKU_PID=655020
+ DOKKU_PID=655020
+ export DOKKU_LOGS_DIR=/var/log/dokku
+ DOKKU_LOGS_DIR=/var/log/dokku
+ export DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ export DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ export DOKKU_CONTAINER_LABEL=dokku
+ DOKKU_CONTAINER_LABEL=dokku
+ export 'DOKKU_GLOBAL_BUILD_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_BUILD_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ export 'DOKKU_GLOBAL_RUN_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_RUN_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ parse_args ps:inspect myapp --trace
+ declare 'desc=top-level cli arg parser'
+ local next_index=1
+ local skip=false
+ args=('ps:inspect' 'myapp' '--trace')
+ local args
+ local flags
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ ps:inspect == \-\-\a\p\p ]]
+ [[ ps:inspect =~ ^--.* ]]
+ next_index=2
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ myapp == \-\-\a\p\p ]]
+ [[ myapp =~ ^--.* ]]
+ next_index=3
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ export DOKKU_TRACE=1
+ DOKKU_TRACE=1
+ [[ false == \t\r\u\e ]]
+ [[ --trace == \-\-\a\p\p ]]
+ [[ --trace =~ ^--.* ]]
+ flags=' --trace'
+ next_index=4
+ [[ -z --trace ]]
+ return 0
+ args=("$@")
+ skip_arg=false
+ [[ ps:inspect =~ ^--.* ]]
+ has_tty
+ declare 'desc=return 0 if we have a tty'
+ [[ '' == \t\r\u\e ]]
+ [[ '' == \t\r\u\e ]]
++ LC_ALL=C
++ /usr/bin/tty
+ [[ /dev/pts/0 == \n\o\t\ \a\ \t\t\y ]]
+ return 0
++ id -un
+ [[ dokku != \d\o\k\k\u ]]
+ [[ ps:inspect =~ ^plugin:.* ]]
+ [[ ps:inspect == \s\s\h\-\k\e\y\s\:\a\d\d ]]
+ [[ ps:inspect == \s\s\h\-\k\e\y\s\:\r\e\m\o\v\e ]]
+ [[ ps:inspect == \s\c\h\e\d\u\l\e\r\-\k\3\s\:\i\n\i\t\i\a\l\i\z\e ]]
+ [[ ps:inspect == \s\c\h\e\d\u\l\e\r\-\k\3\s\:\u\n\i\n\s\t\a\l\l ]]
+ [[ -n '' ]]
+ dokku_auth ps:inspect myapp --trace
+ declare 'desc=calls user-auth plugin trigger'
+ export SSH_USER=root
+ SSH_USER=root
+ export SSH_NAME=default
+ SSH_NAME=default
+ export DOKKU_COMMAND=ps:inspect
+ DOKKU_COMMAND=ps:inspect
++ wc -l
++ find /var/lib/dokku/plugins/enabled/20_events/user-auth
+ local user_auth_count=1
+ [[ 1 == 0 ]]
+ [[ 1 == 1 ]]
+ [[ -f /var/lib/dokku/plugins/enabled/20_events/user-auth ]]
+ return 0
+ case "$1" in
+ execute_dokku_cmd ps:inspect myapp --trace
+ declare 'desc=executes dokku sub-commands'
+ local PLUGIN_NAME=ps:inspect
+ local PLUGIN_CMD=ps:inspect
+ local implemented=0
+ local script
+ argv=('ps:inspect' 'myapp' '--trace')
+ local argv
+ case "$PLUGIN_NAME" in
++ readlink -f /var/lib/dokku/plugins/enabled/ps
+ [[ /var/lib/dokku/core-plugins/available/ps == *core-plugins* ]]
+ [[ ps:inspect == \p\s\:\i\n\s\p\e\c\t ]]
+ shift 1
+ [[ -n '' ]]
+ set -- ps:inspect myapp --trace
+ [[ -x /var/lib/dokku/plugins/enabled/ps:inspect/subcommands/default ]]
+ [[ -x /var/lib/dokku/plugins/enabled/ps:inspect/subcommands/ps:inspect ]]
+ [[ -x /var/lib/dokku/plugins/enabled/ps/subcommands/inspect ]]
+ [[ -n inspect ]]
+ /var/lib/dokku/plugins/enabled/ps/subcommands/inspect ps:inspect myapp --trace
unknown flag: --trace
Usage of ps:inspect:
unknown flag: --trace
```


```
myserver:~# dokku ps:restart myapp --trace
+ export DOKKU_HOST_ROOT=/home/dokku
+ DOKKU_HOST_ROOT=/home/dokku
+ export DOKKU_DISTRO
++ . /etc/os-release
++ echo debian
+ DOKKU_DISTRO=debian
+ export DOCKER_BIN=docker
+ DOCKER_BIN=docker
+ export DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ DOKKU_IMAGE=gliderlabs/herokuish:latest-22
+ export DOKKU_CNB_BUILDER=heroku/builder:22
+ DOKKU_CNB_BUILDER=heroku/builder:22
+ export DOKKU_LIB_ROOT=/var/lib/dokku
+ DOKKU_LIB_ROOT=/var/lib/dokku
+ export PLUGIN_PATH=/var/lib/dokku/plugins
+ PLUGIN_PATH=/var/lib/dokku/plugins
+ export PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ PLUGIN_AVAILABLE_PATH=/var/lib/dokku/plugins/available
+ export PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ PLUGIN_ENABLED_PATH=/var/lib/dokku/plugins/enabled
+ export PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ PLUGIN_CORE_PATH=/var/lib/dokku/core-plugins
+ export PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ PLUGIN_CORE_AVAILABLE_PATH=/var/lib/dokku/core-plugins/available
+ export PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ PLUGIN_CORE_ENABLED_PATH=/var/lib/dokku/core-plugins/enabled
+ export DOKKU_SYSTEM_GROUP=dokku
+ DOKKU_SYSTEM_GROUP=dokku
+ export DOKKU_SYSTEM_USER=dokku
+ DOKKU_SYSTEM_USER=dokku
+ export DOKKU_API_VERSION=1
+ DOKKU_API_VERSION=1
+ export DOKKU_NOT_IMPLEMENTED_EXIT=10
+ DOKKU_NOT_IMPLEMENTED_EXIT=10
+ export DOKKU_VALID_EXIT=0
+ DOKKU_VALID_EXIT=0
+ export DOKKU_PID=654987
+ DOKKU_PID=654987
+ export DOKKU_LOGS_DIR=/var/log/dokku
+ DOKKU_LOGS_DIR=/var/log/dokku
+ export DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ DOKKU_LOGS_HOST_DIR=/var/log/dokku
+ export DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ DOKKU_EVENTS_LOGFILE=/var/log/dokku/events.log
+ export DOKKU_CONTAINER_LABEL=dokku
+ DOKKU_CONTAINER_LABEL=dokku
+ export 'DOKKU_GLOBAL_BUILD_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_BUILD_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ export 'DOKKU_GLOBAL_RUN_ARGS=--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ DOKKU_GLOBAL_RUN_ARGS='--label=org.label-schema.schema-version=1.0 --label=org.label-schema.vendor=dokku --label=dokku'
+ source /var/lib/dokku/core-plugins/available/common/functions
++ set -eo pipefail
++ [[ -n 1 ]]
++ set -x
+ parse_args ps:restart myapp --trace
+ declare 'desc=top-level cli arg parser'
+ local next_index=1
+ local skip=false
+ args=('ps:restart' 'myapp' '--trace')
+ local args
+ local flags
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ ps:restart == \-\-\a\p\p ]]
+ [[ ps:restart =~ ^--.* ]]
+ next_index=2
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ [[ false == \t\r\u\e ]]
+ [[ myapp == \-\-\a\p\p ]]
+ [[ myapp =~ ^--.* ]]
+ next_index=3
+ for arg in "$@"
+ [[ false == \t\r\u\e ]]
+ case "$arg" in
+ export DOKKU_TRACE=1
+ DOKKU_TRACE=1
+ [[ false == \t\r\u\e ]]
+ [[ --trace == \-\-\a\p\p ]]
+ [[ --trace =~ ^--.* ]]
+ flags=' --trace'
+ next_index=4
+ [[ -z --trace ]]
+ return 0
+ args=("$@")
+ skip_arg=false
+ [[ ps:restart =~ ^--.* ]]
+ has_tty
+ declare 'desc=return 0 if we have a tty'
+ [[ '' == \t\r\u\e ]]
+ [[ '' == \t\r\u\e ]]
++ LC_ALL=C
++ /usr/bin/tty
+ [[ /dev/pts/0 == \n\o\t\ \a\ \t\t\y ]]
+ return 0
++ id -un
+ [[ dokku != \d\o\k\k\u ]]
+ [[ ps:restart =~ ^plugin:.* ]]
+ [[ ps:restart == \s\s\h\-\k\e\y\s\:\a\d\d ]]
+ [[ ps:restart == \s\s\h\-\k\e\y\s\:\r\e\m\o\v\e ]]
+ [[ ps:restart == \s\c\h\e\d\u\l\e\r\-\k\3\s\:\i\n\i\t\i\a\l\i\z\e ]]
+ [[ ps:restart == \s\c\h\e\d\u\l\e\r\-\k\3\s\:\u\n\i\n\s\t\a\l\l ]]
+ [[ -n '' ]]
+ dokku_auth ps:restart myapp --trace
+ declare 'desc=calls user-auth plugin trigger'
+ export SSH_USER=root
+ SSH_USER=root
+ export SSH_NAME=default
+ SSH_NAME=default
+ export DOKKU_COMMAND=ps:restart
+ DOKKU_COMMAND=ps:restart
++ wc -l
++ find /var/lib/dokku/plugins/enabled/20_events/user-auth
+ local user_auth_count=1
+ [[ 1 == 0 ]]
+ [[ 1 == 1 ]]
+ [[ -f /var/lib/dokku/plugins/enabled/20_events/user-auth ]]
+ return 0
+ case "$1" in
+ execute_dokku_cmd ps:restart myapp --trace
+ declare 'desc=executes dokku sub-commands'
+ local PLUGIN_NAME=ps:restart
+ local PLUGIN_CMD=ps:restart
+ local implemented=0
+ local script
+ argv=('ps:restart' 'myapp' '--trace')
+ local argv
+ case "$PLUGIN_NAME" in
++ readlink -f /var/lib/dokku/plugins/enabled/ps
+ [[ /var/lib/dokku/core-plugins/available/ps == *core-plugins* ]]
+ [[ ps:restart == \p\s\:\r\e\s\t\a\r\t ]]
+ shift 1
+ [[ -n '' ]]
+ set -- ps:restart myapp --trace
+ [[ -x /var/lib/dokku/plugins/enabled/ps:restart/subcommands/default ]]
+ [[ -x /var/lib/dokku/plugins/enabled/ps:restart/subcommands/ps:restart ]]
+ [[ -x /var/lib/dokku/plugins/enabled/ps/subcommands/restart ]]
+ [[ -n restart ]]
+ /var/lib/dokku/plugins/enabled/ps/subcommands/restart ps:restart myapp --trace
unknown flag: --trace
Usage of ps:restart:
      --all            --all: restart all apps
      --parallel int   --parallel: number of apps to restart in parallel, -1 to match cpu count
unknown flag: --trace
```

**Comment by josegonzalez** [comment]
The trace flag is global, so needs to be set before the subcommand unfortunately.

```
dokku --trace ps:restart myapp
```
